ReactWindows/ChakraBridge/JsIndexedModulesUnbundle.cpp

```diff
@@ -0,0 +1,121 @@
+#include "pch.h"
+#include "ChakraHost.h"
+#include "JsIndexedModulesUnbundle.h"
+
+JsIndexedModulesUnbundle::JsIndexedModulesUnbundle(const wchar_t* sourcePath)
+{
+    this->sourcePath = sourcePath;
+}
+
+JsIndexedModulesUnbundle::~JsIndexedModulesUnbundle()
+{
+	if (fileHandle != NULL)
+	{
+		UnmapViewOfFile(byteBuffer);
+		CloseHandle(mapHandle);
+		CloseHandle(fileHandle);
+	}
+	else
+	{
+		delete[] byteBuffer;
+	}
+}
+
+JsModulesUnbundleModule* JsIndexedModulesUnbundle::GetModule(uint32_t index)
+{
+	ModuleData moduleData = moduleTable.modules[index];
+
+	// Create the source URL
+	auto filename = new wchar_t[_MAX_FNAME];
+	_itow_s(index, filename, _MAX_FNAME, 10);
+	wcscat_s(filename, _MAX_FNAME, L".js");
+
+	// Copy the script
+	auto script = new char[moduleData.length];
+	auto wcScript = new wchar_t[moduleData.length];
+	CopyMemory(script, byteBuffer + baseOffset + moduleData.offset, moduleData.length);
+	size_t ignored;
+	mbstowcs_s(&ignored, wcScript, moduleData.length, script, moduleData.length);
+	delete[] script;
+
+	// Return the module
+	auto module = new JsModulesUnbundleModule();
+	module->source = wcScript;
+	module->sourceUrl = filename;
+	return module;
+}
+
+wchar_t* JsIndexedModulesUnbundle::GetStartupCode()
+{
+	if (ChakraHost::LoadByteCode(this->sourcePath, &byteBuffer, &fileHandle, &mapHandle) == JsError)
+	{
+		return nullptr;
+	}
+
+	// Copy the header data
+	uint32_t header[3];
+	CopyMemory(header, byteBuffer, sizeof(header));
+
+	// Create the module table
+	uint32_t numberOfTableEntries = header[1];
+	moduleTable = ModuleTable(numberOfTableEntries);
+	uint32_t moduleTableSize = moduleTable.byteLength();
+	CopyMemory(moduleTable.modules, byteBuffer + sizeof(header), moduleTableSize);
```

Comment:
Good suggestion here.

---
In reply to: [103548272](https://github.com/ReactWindows/react-native-windows/pull/1017#discussion_r103548272) [](ancestors = 103548272)